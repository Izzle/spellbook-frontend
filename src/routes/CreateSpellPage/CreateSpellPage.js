import React, { Component } from 'react';
import config from '../../config';
import SpellContext from '../../contexts/SpellContext';
import ValidationError from '../../components/ValidationError/ValidationError';

export default class CreateSpellPage extends Component {
  constructor(props) {
    super(props);

    this.state = { 
      name: {
        value: '',
        touched: false
      },
      school: {
        value: '',
        touched: false
      },
      level: {
        value: '',
        touched: false
      },
      castingTime: {
        value: '',
        touched: false
      },
      range: {
        value: '',
        touched: false
      },
      spellComponents: {
        value: '',
        touched: false
      },
      duration: {
        value: '',
        touched: false
      },
      description: {
        value: '',
        touched: false
      },
      higherLevels: {
        value: '',
        touched: false
      }
    }
  }

  static contextType = SpellContext;

  // A form to capture the information for a new spell. Submit to a POST to '/PLACEHOLDER' endpoint

  updateSpell(value, keyName) {
    // Whenever the user types anything into the input, we update the state! each letter at a time!
    // When the user types ANYTHING we set the value of touched to "true" from "false" this way we know
    // when to correctly render our ValidationError
    let newState = { [keyName]: {value, touched: true}};
    this.setState(newState);
  }

  validateLevel() {
    const level = this.state.level.value.trim();
    if(level.length === 0) {
      return 'Please select a spell level'
    }
  }

  validateName() {
    // Validates the name of a new spell
    // this removes any whitespace from value and makes sure the value isnt empty
    const name = this.state.name.value.trim();
    if(name.length === 0) {
      return 'Please enter a spell name';
    } else if(name.length > 30) {
      return 'Please keep the spell name under 30 characters'
    }
  }

  handleSubmit(event) {
    event.preventDefault();
    // object destructing to get our variables from state
    const {
      name,
      school,
      level,
      castingTime,
      range,
      spellComponents,
      duration,
      description,
      higherLevels
    } = this.state;
    
    // formatting our data so the server will receive the correct variable names in our POST request
    const data = {
      spell_name: name.value,
      spell_school: school.value,
      spell_level: parseInt(level.value),
      cast_time: castingTime.value,
      spell_range: range.value,
      spell_components: spellComponents.value,
      spell_duration: duration.value,
      spell_description: description.value,
      higher_levels: higherLevels.value
    };

    fetch(`${config.API_ENDPOINT}/spells`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(data)
    })
    .then(res => {
      return (!res.ok)
                ? res.json().then(e => Promise.reject(e))
                : res.json();
    })
    .then(this.context.addSpell)
    .catch(this.context.setError);
    // REFACTOR THIS TO USE 'SpellApiService.postSpell(data);

// SpellApiService.getAllSpells()
//        .then(this.context.setSpells)
//        .catch(this.context.setError);

  }
  

  render() {
    const levelError = this.validateLevel();
    const nameError = this.validateName();

    return (
      <section className='CreateSpellPage__main'>
          <form className='CreateSpellPage__form' onSubmit={e => this.handleSubmit(e)}>
            <h2>Create a new Spell</h2>
            <div className='form-group'>
              <label htmlFor='add-spell-name'>Spell Name </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-name'
                id='add-spell-name'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'name')}
                required
                />
              {/* Conditional rendering depending on whether the user has changed the input or not */}
              {this.state.name.touched && (<ValidationError message={nameError} />)}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-school'>Magic School </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-school'
                id='add-spell-school'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'school')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className='form-group'>
              <label htmlFor="add-spell-level">Level </label>
              <select 
                name="add-spell-level"
                id="add-spell-level"
                onChange={e => this.updateSpell(e.target.value, 'level')}
                required>
                  <option value="">Select spell level</option>
                  <option value="0" defaultValue="selected">Cantrip</option>
                  <option value="1">1</option>
                  <option value="2">2</option>
                  <option value="3">3</option>
                  <option value="4">4</option>
                  <option value="5">5</option>
                  <option value="6">6</option>
                  <option value="7">7</option>
                  <option value="8">8</option>
                  <option value="9">9</option>
              </select>
              {this.state.level.touched && (<ValidationError message={levelError} />)}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-castingTime'>Casting Time </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-castingTime'
                id='add-spell-castingTime'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'castingTime')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-range'>Range </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-range'
                id='add-spell-range'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'range')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-components'>Components </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-components'
                id='add-spell-components'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'spellComponents')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-duration'>Duration </label>
              <input
                type='text'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-duration'
                id='add-spell-duration'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'duration')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className='form-group'>
              <label htmlFor='add-spell-description'>Description </label>
              <textarea
                rows='5'
                cols='35'
                className='CreateSpellPage__control'
                // on submit, the name will what the server will get data keys name after
                name='add-spell-description'
                id='add-spell-description'
                 // Each time the user types anything into input, we update the state. This allows real-time error messages (ie before submit)
                onChange={e => this.updateSpell(e.target.value, 'description')}
                required
                />
                {/* ADD VALIDATION HERE */}
            </div>

            <div className="CreateSpellPage__button__group">
              <button type="reset" className="CreateSpellPage__button">Cancel</button>
              <button
                type="submit"
                className="CreateSpellPage__button"
                 // if anything triggers our validation method, the button gets disabled
                disabled={this.validateLevel() || this.validateName()}>Create Spell</button>
            </div>
          </form>
      </section>
    )
  }
}