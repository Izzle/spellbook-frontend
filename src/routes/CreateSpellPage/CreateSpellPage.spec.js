import React from 'react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import CreateSpellPage from './CreateSpellPage';
import { BrowserRouter } from 'react-router-dom';


describe('<CreateSpellPage />', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<BrowserRouter><CreateSpellPage /></BrowserRouter>, div);
        ReactDOM.unmountComponentAtNode(div);
    });
    it('renders this UI as expected', () => {
        const tree = renderer.create(<BrowserRouter><CreateSpellPage /></BrowserRouter>).toJSON();
        expect(tree).toMatchSnapshot();
    });    
});