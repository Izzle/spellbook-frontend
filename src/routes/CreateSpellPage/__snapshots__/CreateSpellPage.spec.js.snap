// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateSpellPage /> renders this UI as expected 1`] = `
<section
  className="CreateSpellPage__main"
>
  <form
    className="CreateSpellPage__form"
    onSubmit={[Function]}
  >
    <h2>
      Create a New Spell
    </h2>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-name"
      >
        Spell Name 
      </label>
      <input
        className="CreateSpellPage__control"
        id="add-spell-name"
        name="add-spell-name"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-school"
      >
        Magic School 
      </label>
      <select
        id="add-spell-school"
        name="add-spell-school"
        onChange={[Function]}
        required={true}
      >
        <option
          value=""
        >
          Select Magic School
        </option>
        <option
          defaultValue="selected"
          value="Conjuration"
        >
          Conjuration
        </option>
        <option
          value="Necromancy"
        >
          Necromancy
        </option>
        <option
          value="Evocation"
        >
          Evocation
        </option>
        <option
          value="Abjuration"
        >
          Abjuration
        </option>
        <option
          value="Transmutation"
        >
          Transmutation
        </option>
        <option
          value="Divination"
        >
          Divination
        </option>
        <option
          value="Enchantment"
        >
          Enchantment
        </option>
        <option
          value="Illusion"
        >
          Illusion
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-level"
      >
        Level 
      </label>
      <select
        id="add-spell-level"
        name="add-spell-level"
        onChange={[Function]}
        required={true}
      >
        <option
          value=""
        >
          Select spell level
        </option>
        <option
          defaultValue="selected"
          value="0"
        >
          Cantrip
        </option>
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="7"
        >
          7
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="9"
        >
          9
        </option>
      </select>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-castingTime"
      >
        Casting Time 
      </label>
      <input
        className="CreateSpellPage__control"
        id="add-spell-castingTime"
        name="add-spell-castingTime"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-range"
      >
        Range 
      </label>
      <input
        className="CreateSpellPage__control"
        id="add-spell-range"
        name="add-spell-range"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-components"
      >
        Components 
      </label>
      <input
        className="CreateSpellPage__control"
        id="add-spell-components"
        name="add-spell-components"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-duration"
      >
        Duration 
      </label>
      <input
        className="CreateSpellPage__control"
        id="add-spell-duration"
        name="add-spell-duration"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="add-spell-description"
      >
        Description 
      </label>
      <textarea
        className="CreateSpellPage__control"
        cols="35"
        id="add-spell-description"
        maxLength={2500}
        name="add-spell-description"
        onChange={[Function]}
        required={true}
        rows="5"
      />
    </div>
    <div
      className="CreateSpellPage__button__group"
    >
      <button
        className="CreateSpellPage__button"
        type="reset"
      >
        Cancel
      </button>
      <button
        className="CreateSpellPage__button"
        disabled="Please select a spell level"
        type="submit"
      >
        Create Spell
      </button>
    </div>
  </form>
</section>
`;
